pandas/core/frame.py

```diff
@@ -650,25 +650,36 @@ def _constructor(self) -> type[DataFrame]:
         return DataFrame
 
     def _constructor_from_mgr(self, mgr, axes) -> DataFrame:
-        if self._constructor is DataFrame:
-            # we are pandas.DataFrame (or a subclass that doesn't override _constructor)
-            return DataFrame._from_mgr(mgr, axes=axes)
-        else:
-            assert axes is mgr.axes
+        df = DataFrame._from_mgr(mgr, axes=axes)
+
+        if type(self) is DataFrame:
+            # This would also work `if self._constructor is DataFrame`, but
+            #  this check is slightly faster, benefiting the most-common case.
+            return df
+
+        elif type(self).__name__ == "GeoDataFrame":
```

Comment:
Worth having a test for this in `test_downstream.py`? `geopandas` is installed in the CI

Reply:
Good idea (also should add a test for#57032).  Will wait until Joris weighs in

Reply:
im having trouble writing a meaningful test for this cc @jorisvandenbossche

Reply:
GeoPandas is already overriding `_constructor_from_mgr`, so normally this shouldn't be needed.

(will try to take a closer look one of the coming days!)